// @apollo/gateway/dist/loadServicesFromRemoteEndpoint.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const node_fetch_1 = require("node-fetch");
const _1 = require("./");
function getServiceDefinitionsFromRemoteEndpoint({ serviceList, headers = {}, serviceSdlCache, }) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!serviceList || !serviceList.length) {
            throw new Error('Tried to load services from remote endpoints but none provided');
        }
        let isNewSchema = false;
        const promiseOfServiceList = serviceList.map(({ name, url, dataSource }) => {
            if (!url) {
                throw new Error(`Tried to load schema for '${name}' but no 'url' was specified.`);
            }
            const request = {
                query: _1.SERVICE_DEFINITION_QUERY,
                http: {
                    url,
                    method: 'POST',
                    headers: new node_fetch_1.Headers(headers),
                },
            };
            return dataSource
                .process({ request, context: {} })
                .then(({ data, errors }) => {
                if (data && !errors) {
                    const typeDefs = data._service.sdl;
                    const previousDefinition = serviceSdlCache.get(name);
                    if (previousDefinition !== typeDefs) {
                        isNewSchema = true;
                    }
                    serviceSdlCache.set(name, typeDefs);
                    return {
                        name,
                        url,
                        typeDefs: graphql_1.parse(typeDefs),
                    };
                }
                throw new Error(errors === null || errors === void 0 ? void 0 : errors.map(e => e.message).join("\n"));
            })
                .catch(err => {
                const errorMessage = `Couldn't load service definitions for "${name}" at ${url}` +
                    (err && err.message ? ": " + err.message || err : "");
                throw new Error(errorMessage);
            });
        });
        const serviceDefinitions = yield Promise.all(promiseOfServiceList);
        return { serviceDefinitions, isNewSchema };
    });
}
exports.getServiceDefinitionsFromRemoteEndpoint = getServiceDefinitionsFromRemoteEndpoint;
